speechd: 0.11.5 -> 0.12.1

pitkling opened the pull request:
The upgrade to `0.12.0-rc4` enables building `speechd-minimal` on Darwin (see the [upstream changelog](https://github.com/brailcom/speechd/releases/tag/0.12.0-rc4)).


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pitkling committed:
speechd: 0.11.5 → 0.12.1

https://github.com/brailcom/speechd/releases/tag/0.12.0
https://github.com/brailcom/speechd/releases/tag/0.12.1

The upgrade to `0.12.0` enables building `speechd-minimal` on Darwin.

jtojnar commented on pkgs/by-name/sp/speechd/package.nix:
Thanks, module should be updated as well:

https://github.com/NixOS/nixpkgs/blob/84ed163a938bf27ba6839667444fc2b9a134aa2e/nixos/modules/services/accessibility/speechd.nix#L24

pitkling replied:
Ah, I didn't look into the module code, since I require `speechd` merely as a dependency of another package.

Not quite sure how to switch to socket activation here as requested by the comment. I'll have to take a look at the `speechd` docs first (will probably take a bit until I get to it).

pitkling replied:
Thanks, I added your suggested change. I haven't tested it yet, since I don't have a Linux machine at hand right now. I'll leave a comment once I do.

pitkling committed:
speechd: use socket activation

jtojnar commented on pkgs/by-name/sp/speechd/package.nix:
```diff
@@ -77,9 +77,11 @@ stdenv.mkDerivation rec {
       libpulseaudio
-      alsa-lib
       python
     ]
```
Looks like it also needs `libsystemd` (`systemdMinimal` package should probably work).

jtojnar commented on pkgs/by-name/sp/speechd/package.nix:
`fix-paths.patch` should be adjusted to also patch https://github.com/brailcom/speechd/blob/2d9069d74fa72fcdad228e07b28cf6b483f5de96/speech-dispatcher.service.in#L23

jtojnar reviewed (commented):
I went through the upstream diffs:

- https://github.com/brailcom/speechd/compare/0.11.5...0.12.0-rc1
- https://github.com/brailcom/speechd/compare/0.12.0-rc1...0.12.0-rc2
- https://github.com/brailcom/speechd/compare/0.12.0-rc2...0.12.0-rc3
- https://github.com/brailcom/speechd/compare/0.12.0-rc3...0.12.0-rc4

The only other significant changes I noticed are:

- addition of PipeWire support – probably not needed for now, libpulse is still recommended for clients by PipeWire IIRC
- addition of [Open JTalk](https://open-jtalk.sourceforge.net/) module – not currently packaged in Nixpkgs

pitkling commented:
@jtojnar Thanks for the help! I did the updates as you suggested, but cannot test them myself before tomorrow.

pitkling commented on pkgs/by-name/sp/speechd/package.nix:
```diff
@@ -146,7 +153,7 @@ stdenv.mkDerivation rec {
       jtojnar
     ];
-    platforms = platforms.linux;
+    platforms = platforms.unix;
```
@jtojnar A quick question. Technically, the update only enables the compilation of `speechd-minimal` on Darwin at the moment. However, I have pending PRs (#375450 and #375038) that update the `speechd` dependencies `svox` and `mbrola` such that they compile on Darwin. With those, `speechd` also compiles on Darwin.

So maybe for now this should remain `platforms.linux` and I could add a corresponding overwrite of `meta.platforms` to `speechd-minimal`. What do you think?

jtojnar replied:
We do sometimes use conditionals based on features, that will probably be the cleanest. Example:

https://github.com/NixOS/nixpkgs/blob/3c6c229eb23c75e38cbd2b7740d75d77835411d3/pkgs/by-name/vi/vifm/package.nix#L73

pitkling replied:
That makes sense, thanks! Done.

pitkling commented:
@jtojnar: I tested the socket activation on Linux, which seems to work now (I had to add the configure flag `--with-systemduserunitdir=${placeholder "out"}/lib/systemd/user`).

I also made the svox/pico-related  `postPatch` conditional on `withPico` to avoid the `svox` dependency when building `speechd-minimal`.

pitkling commented:
Question for someone more experienced with nixpkgs than me (@jtojnar?): Should this be rebased onto staging, given that it causes >500 rebuilds (at least on Linux; on Darwin it's much less)?

Also, is there anything else I can do to increase the chance to get this reviewed & eventually merged?

pitkling commented:
Upstream recently released 0.12.0, so I updated the PR accordingly (and removed a check that's no longer needed since PR #375038 got merged). I had a quick look at the [upstream diff between 0.12.0-rc4 and 0.12.0](https://github.com/brailcom/speechd/compare/0.12.0-rc4...0.12.0). It looks fine to me. Tested compilation again both on `aarch64-darwin` and `x86_64-linux`.

@jtojnar: Could you take another look at this? Would love to get this merged. :slightly_smiling_face:

jtojnar reviewed (changes requested):
Thanks. I reviewed the new upstream changes and they look okay as well: https://github.com/brailcom/speechd/compare/0.12.0-rc4...0.12.0

I also finally gotten around to testing it. Unfortunately, it appears that it causes a regression compared to `HEAD^^^` with Orca in GNOME VM. The following is logged into the console upon enabling _Screen Reader_ in the _accessibility menu_:

```
Mar 01 15:16:01 nixos gst-plugin-scan[2602]: pygobject initialization failed: could not import gobject (error was: ModuleNotFoundError("No module named 'gi'"))
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: Traceback (most recent call last):
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 96, in init
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _init_speech_server(module_name, settings.speechServerInfo)
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 77, in _init_speech_server
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     raise RuntimeError(f"ERROR: No speech server for factory: {module_name}")
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: RuntimeError: ERROR: No speech server for factory: speechdispatcherfactory
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: During handling of the above exception, another exception occurred:
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: Traceback (most recent call last):
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 102, in init
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _init_speech_server(module_name, None)
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 71, in _init_speech_server
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _speechserver = factory.SpeechServer.get_speech_server()
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: TypeError: SpeechServer.get_speech_server() missing 1 required positional argument: 'info'
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: Traceback (most recent call last):
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 96, in init
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _init_speech_server(module_name, settings.speechServerInfo)
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 77, in _init_speech_server
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     raise RuntimeError(f"ERROR: No speech server for factory: {module_name}")
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: RuntimeError: ERROR: No speech server for factory: speechdispatcherfactory
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: During handling of the above exception, another exception occurred:
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: Traceback (most recent call last):
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 102, in init
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _init_speech_server(module_name, None)
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 71, in _init_speech_server
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _speechserver = factory.SpeechServer.get_speech_server()
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: TypeError: SpeechServer.get_speech_server() missing 1 required positional argument: 'info'
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: Traceback (most recent call last):
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 96, in init
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _init_speech_server(module_name, settings.speechServerInfo)
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 77, in _init_speech_server
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     raise RuntimeError(f"ERROR: No speech server for factory: {module_name}")
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: RuntimeError: ERROR: No speech server for factory: speechdispatcherfactory
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: During handling of the above exception, another exception occurred:
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: Traceback (most recent call last):
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 102, in init
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _init_speech_server(module_name, None)
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 71, in _init_speech_server
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _speechserver = factory.SpeechServer.get_speech_server()
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: TypeError: SpeechServer.get_speech_server() missing 1 required positional argument: 'info'
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: Traceback (most recent call last):
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 96, in init
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _init_speech_server(module_name, settings.speechServerInfo)
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 77, in _init_speech_server
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     raise RuntimeError(f"ERROR: No speech server for factory: {module_name}")
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: RuntimeError: ERROR: No speech server for factory: speechdispatcherfactory
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: During handling of the above exception, another exception occurred:
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: Traceback (most recent call last):
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 102, in init
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _init_speech_server(module_name, None)
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:   File "/nix/store/pr9yjcpjzlirzkmx7iijp90kz9fcv0cs-orca-47.3/lib/python3.12/site-packages/orca/speech.py", line 71, in _init_speech_server
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:     _speechserver = factory.SpeechServer.get_speech_server()
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]:                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 01 15:16:02 nixos orca-autostart.desktop[2572]: TypeError: SpeechServer.get_speech_server() missing 1 required positional argument: 'info'
```

pitkling commented:
> […] I also finally gotten around to testing it. Unfortunately, it appears that it causes a regression compared to `HEAD^^^` with Orca in GNOME VM. The following is logged into the console upon enabling _Screen Reader_ in the _accessibility menu_: […]

Thanks for testing it. I'll take a look at it later, but probably only sometime later this week.

pitkling commented:
> Thanks. I reviewed the new upstream changes and they look okay as well: [brailcom/speechd@0.12.0-rc4...0.12.0](https://github.com/brailcom/speechd/compare/0.12.0-rc4...0.12.0)
> 
> I also finally gotten around to testing it. Unfortunately, it appears that it causes a regression compared to `HEAD^^^` with Orca in GNOME VM. The following is logged into the console upon enabling _Screen Reader_ in the _accessibility menu_: […]

I finally got around to this @jtojnar. I reproduced the error message by running `orca -s` on my Linux machine (which pulls in the updated `speechd` package/module from this PR) and then used `orca -s --debug` to see what's going on.

I'm not too familiar with `systemd` and how it is handled in NixOS (I'm mostly on Darwin), but the problem seems to be that the [socket user service](https://github.com/brailcom/speechd/blob/0.12.0/speech-dispatcher.socket) is not enabled. Indeed, after manually starting it via `systemctl start --user speech-dispatcher.socket`, the above `orca -s` works. I was originally under the impression that adding `speechd` to `systemd.packages` would enable the user service, but that [seems not to be true](https://nixos.org/manual/nixos/stable/#sect-nixos-systemd-nixos) (at least not for user services)?

@jtojnar: How do NixOS modules usually enable upstream user service files added via `systemd.packages`? I assume something like [this example](https://github.com/NixOS/nixpkgs/blob/4f48368f11e7329735ab76d890f18f8d4be3f60f/nixos/modules/services/web-servers/static-web-server.nix#L55-L63)? I tested this locally by adding
```nix
systemd.user.sockets.speech-dispatcher.wantedBy = [ "sockets.target" ];
```
to the `speechd` NixOS module and it works for me.

pitkling commented:
@jtojnar: Just a friendly reminder to check whether my update fixed your Orca issue, in case it slipped through your inbox. Don't worry if you still do not have time to recheck.

nyabinary commented:
[speechd 12.1 is out](https://github.com/brailcom/speechd/releases/tag/0.12.1)

pitkling commented:
> [speechd 12.1 is out](https://github.com/brailcom/speechd/releases/tag/0.12.1)

Thanks for the heads-up. From a (very) [quick look](https://github.com/brailcom/speechd/compare/0.12.0...0.12.1) it seems there are no potential problematic updates between 0.12.0 and 0.12.1. I'll update the PR accordingly. But it'll take a few days until I have time for that…

jtojnar commented:
Thanks. Your analysis makes sense to me. We have an issue about the `[Install]` section: https://github.com/NixOS/nixpkgs/issues/81138

I bumped it to .1 and tested with GNOME 48.
